### PR TITLE
Input Preparation & Routing

### PR DESCRIPTION
**Pull request by styagi130**
# What does this PR do ?

Batch EasyMagpie input preparation and feedback snapshots, move supported codec compilation into startup, and remove a redundant conversion in the eligible backbone router. Preserve request ownership, routing calculations and fallback behavior.

**Collection**: TTS (`easymagpie_vllm_omni`).

## Scope against main

This is MR 4 of 5. Its branch includes the following prerequisite changes in the current diff to `main`:

- **Serving & Benchmarks** (`serving-benchmarks`).
- **Decoding & Audio Transfers** (`decoding-audio-transfers`).
- **Voice Preloading** (`voice-preloading`).

The changelog below describes the additional work in this layer. The full current comparison is **12 commits; 52 files changed, 7745 insertions(+), 518 deletions(-)**. Merge the preceding MRs first. If their commits are preserved, those changes drop out of this diff as `main` advances. After a squash or rebase, restack and revalidate this branch before merging.

# Changelog

- Batch compatible text/phoneme preparation and use validated request spans for fixed decode-prefix dispatch, retaining scalar and dynamic fallbacks.
- Take owning feedback snapshots before asynchronous output while keeping immediate stores where preparation order requires them.
- Derive eligible Mamba prefill state flags from equivalent CPU metadata, with guarded installation and unsupported-path fallbacks.
- Pair contiguous codec weight layout with matching kernel addressing and IEEE accumulation; warm supported packed paths using temporary state pages before readiness.
- Remove the redundant FP16-to-FP32 router-logits copy only for eligible unquantized grouped-router configurations. Change gate output and router input dtypes together while keeping FP32 routing calculations.
- Add metadata, ownership, warmup, kernel, expert-selection and graph-replay regressions, and document activation and startup costs.

# Usage

Eligible paths activate automatically with the pinned runtime. Codec warmup uses the codec generation worker from the decoding branch and adds startup work. Existing clients and supported voices keep their interfaces; unsupported metadata, layouts and router configurations retain their fallback paths.

# GitHub Actions CI

Remote CI is pending. Trusted PRs use the existing copy-pr-bot flow. For an untrusted PR, a maintainer can trigger CI with `/ok to test <head-sha>` after publication.

## Validation

Recorded at `1b2a594c5402c082a66aa7ad66052cbfc1f65154` on **2026-09-15**: **984 tests passed, zero failures and zero skips** in the complete CPU/CUDA serving suite. Tests used physical **H100 device 1**, exposed as the sole logical CUDA device in the container.

The package was freshly installed from this head. Source hashes, installed module origins, `pip check`, dependency invariance and author-matching DCO sign-offs passed. Full-branch pre-commit passed from the recorded main base. The runtime used vLLM/vLLM-Omni 0.26.0, PyTorch 2.11.0+cu130 and Triton 3.6.0.

<details>
<summary>Recorded validation commands</summary>

The serving command ran inside the pinned test container after installing this branch and providing the codec contract fixture:

```bash
/usr/bin/python3 -I -B /recipe/installed_pytest.py \
  --rootdir=/snapshot \
  --confcutdir=/snapshot/tests/collections/tts/easymagpie_vllm_omni/serving \
  /snapshot/tests/collections/tts/easymagpie_vllm_omni/serving \
  -p no:cacheprovider -q -ra --junitxml=/results/tests.xml \
  --durations=10
```

Full-branch formatting checks and the current diff check:

```bash
python3 -m pre_commit run \
  --from-ref ea1ebf55b25fceff372f885b4e7b5803f3d33dea \
  --to-ref 1b2a594c5402c082a66aa7ad66052cbfc1f65154
git diff --check ea1ebf55b25fceff372f885b4e7b5803f3d33dea 1b2a594c5402c082a66aa7ad66052cbfc1f65154
```

</details>

Live HTTP/WebSocket audio serving, cancellation/recovery, shutdown/drain, full-utterance quality/completion, performance acceptance and remote CI remain pending. The local test counts do not establish benchmark throughput or latency.

# Before your PR is "Ready for review"

- [x] Add regression tests and documentation with the implementation.
- [x] Pass the complete installed CPU/CUDA serving suite at the recorded head.
- [x] Pass full-branch pre-commit, diff checks and author-matching DCO checks.
- [ ] Complete contributor-guideline review and reviewer assessment of tests, documentation and optional dependency guards.
- [ ] Complete the remaining live serving, quality/completion and performance checks; attach portable validation evidence and pass remote CI.

**PR Type**: New Feature.

## Who can review?

TTS serving, codec and backbone/router maintainers.

# Additional Information

Related to NVIDIA-NeMo/Speech#16220. Merge order: `serving-benchmarks` → `decoding-audio-transfers` → `voice-preloading` → `input-preparation-routing` → `h100-deployment`.

> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

Trusted PRs run automatically through copy-pr-bot. For an untrusted PR, a maintainer can trigger CI by commenting
`/ok to test <head-sha>`; repeat this after a new push if the PR remains untrusted.

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Speech/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA-NeMo/Speech/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: tools/easymagpie_vllm_omni/README.md [hunk 1]
```diff
@@ -93,6 +93,19 @@ keep their previous values. Loading uses `weights_only=True` and checks shape, f
 before device transfer. These limits bound retained startup state, not peak deserialization memory. This moves
 cold voice work into startup without changing formats, default context, or promising warm throughput gains.
 
+### Batch preparation and startup compilation
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/README.md:96]: drop this from toplevel README. those are implementation details

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/codec/kernels.py [hunk 2]
```diff
@@ -459,8 +462,8 @@ def packed_causal_conv1d(
     if not inputs.is_cuda:
         raise ValueError("packed_causal_conv1d is a CUDA kernel")
     inputs = inputs.contiguous()
-    weight = weight.contiguous()
     output_channels, input_channels, kernel_size = weight.shape
+    weight = weight.permute(2, 1, 0).contiguous()
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/codec/kernels.py:466]: if we need weight in different shape, cant we declare it and do it once in model conversion step?

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/backbone_patches.py [hunk 3]
```diff
@@ -14,14 +14,73 @@
...
 import torch
 import vllm.v1.attention.backends.mamba_attn as _mamba_attn
 from vllm.logger import init_logger
 from vllm.model_executor.layers.activation import ReLUSquaredActivation, get_act_fn
+from vllm.v1.attention.backends.mamba2_attn import Mamba2AttentionMetadataBuilder
 
 logger = init_logger(__name__)
 
 
+def patch_mamba_prefill_initial_states() -> None:
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/backbone_patches.py:29]: here we patch `Mamba2AttentionMetadataBuilder.build`. original builder computes prep_initial_states by reducing has_initial_states_p. patch replaces it with equivalent metadata.

Unless there is strong evidence that it improves performance (throughput / latency) for different concurrencies, i would be reluctant to patch the upstream like this. 

and if there is indication that this is helpful - we need to create PR to upstream.

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/easymagpie.py [hunk 4]
```diff
@@ -1061,23 +1220,29 @@ def _preprocess_decode(
         # stability but its codes for those frames are discarded by the caller and
         # never fed back here.
         if decode_offset < self.speech_delay:
-            self._dec_audio_valid[start] = 0
+            self._dec_audio_valid[start].fill_(0)
         elif decode_offset == self.speech_delay:
             self._dec_audio_codes[start].fill_(self.arch.audio_bos_id)
-            self._dec_audio_valid[start] = 1
+            self._dec_audio_valid[start].fill_(1)
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/easymagpie.py:1226]: using "fill" instead of assignment - reasonable standalone microoptimization. avoids creating / copying a tensor.

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/easymagpie.py [hunk 5]
```diff
@@ -603,6 +636,103 @@ def make_omni_output(self, model_outputs, **_: Any) -> OmniOutput:
...
+        if prefix is None:
+            prefix = ()
+        if not isinstance(prefix, (list, tuple)) or len(prefix) > count:
+            return None
+        if any(type(token) is not int or not 0 <= token < self.text_embedding.num_embeddings for token in prefix):
+            return None
+        return count, tuple(prefix)
+
+    def _prepare_batch_decode(self, infos, device) -> None:
+        """Batch feedback calculations without choosing request rows or phases."""
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/easymagpie.py:699]: thats specifically phoneme feedback calculations.

again data-driven justification needed.

if that is general problem - we need to review how can we do batched processing for preprocess/postprocess in general.

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/runner.py [hunk 6]
```diff
@@ -165,6 +242,10 @@ def sample_tokens(self, grammar_output=None):
 class EasyMagpieCodecGPUGenerationWorker(GPUGenerationWorker):
     """Construct the codec runner through the upstream worker."""
 
+    def compile_or_warm_up_model(self):
+        self.get_model().warmup_codec()
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/runner.py:246]: i am concerned that we dont have a very clear idea what to warm up for. 
it depends on usecase (concurrency, arrival time).

please show that this is helpful or lets not additionally complicate the pipeline.

File: tools/easymagpie_vllm_omni/easymagpie_vllm_omni/easymagpie.py [hunk 7]
```diff
@@ -352,7 +356,27 @@ def _select_query_layout(attn_metadata):
                 break
         return None, None
 
-    def _get_query_dispatch(self):
+    @staticmethod
+    def _leading_decode_count(spans, num_queries, max_query_len):
```
Comment (vklimkov-nvidia) [inline tools/easymagpie_vllm_omni/easymagpie_vllm_omni/easymagpie.py:360]: is this python loop really needed?
you can calculate it once in the runner.

**Review by vklimkov-nvidia — COMMENTED** [review]
PR contains some meaningful and potentially impactful optimizations, although it is to be proved.

* optimizing mamba state handling - lets upstream if really helps.
* lets introduce postprocess_batch to mirror preprocess_batch for vllm-omni. this will take away a lot of ad-hoc batching introduced here.
* codec triton kernel improvement can be valuable. can be tweaked a bit more and benchmark needed.
* avoiding dynamic cuda indexing - could help too, commit can be additionally cleaned from extra python loop introduced.
* router improvement - again to verify and upstream.
